jax/interpreters/pxla.py

```diff
@@ -663,24 +663,34 @@ def parallel_callable(fun, backend, axis_name, axis_size, global_axis_size,
   inner_pmap = len(_thread_local_state.dynamic_axis_env) > 0
 
   # Determine global_axis_size for use in AxisEnv.
-  must_run_on_all_devices = True
-  if devices:
-    assert global_axis_size is None  # Checked in api.py
-    global_axis_size = len(devices)
-    must_run_on_all_devices = False
-  elif xb.host_count() > 1:
-    if global_axis_size is None:
-      if inner_pmap:
-        raise ValueError("'axis_size' must be specified for nested multi-host pmaps")
-      global_axis_size = axis_size * xb.host_count()
-  else:
-    if global_axis_size is not None:
-      if global_axis_size != axis_size:
-        raise ValueError(
-            "Specified axis_size {} doesn't match received axis_size {}.".format(
-                global_axis_size, axis_size))
-    else:
+  if xb.host_count() > 1 and global_axis_size is None and inner_pmap:
+    raise ValueError("'axis_size' must be specified for nested multi-host pmaps")
+  if (xb.host_count() == 1 and global_axis_size is not None and
+      global_axis_size != axis_size):
+    raise ValueError(
+        f"Specified axis_size {global_axis_size} doesn't match received "
+        f"axis_size {axis_size}.")
+
+  must_run_on_all_devices = False
+  no_nested_sharding = False
+  if global_axis_size is None:
+    if xb.host_count() == 1:
       global_axis_size = axis_size
+    elif devices:
+      # This allows each host in a multi-host pmap to run on a different number
+      # of devices, but precludes nested sharding (i.e. inner pmaps or
+      # sharded_jits).
+      global_axis_size = len(devices)
+      no_nested_sharding = True
+    else:
+      # This assumes all hosts run on the same number of devices. We make sure
+      # this assumption is true by requiring that the pmap is run on all devices
```

Comment:
This is probably fine for now, but there are definitely use cases for e.g. running a pmap on half of the devices on each host where you might still want to use nesting.

Reply:
You can do that, but you have to manually specify `axis_size` in addition to `devices` (note you only get here if global_axis_size is None and multi-host and `devices` isn't specified). There's an error message below recommending to specify `axis_size`. This is extremely difficult to reason about though, I'd like to somehow simplify it but I'm not sure how.